hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieCompactor.java

```diff
@@ -265,11 +267,36 @@ private int doCompact(JavaSparkContext jsc) throws Exception {
           cfg.compactionInstantTime = firstCompactionInstant.get().getTimestamp();
           LOG.info("Found the earliest scheduled compaction instant which will be executed: "
               + cfg.compactionInstantTime);
-        } else {
-          LOG.info("There is no scheduled compaction in the table.");
-          return 0;
+        }
+
+        // update cfg.compactionInstantTime if finding an expired instant
+        List<HoodieInstant> inflightInstants = metaClient.getActiveTimeline()
+            .filterPendingCompactionTimeline().filterInflights().getInstants();
+        List<String> expiredInstants = inflightInstants.stream().filter(instant -> {
+          try {
+            return client.getHeartbeatClient().isHeartbeatExpired(instant.getTimestamp());
+          } catch (IOException io) {
+            LOG.info("Failed to check heartbeat for instant " + instant);
+          }
+          return false;
+        }).map(HoodieInstant::getTimestamp).collect(Collectors.toList());
+
+        if (!expiredInstants.isEmpty()) {
+          cfg.compactionInstantTime = expiredInstants.get(0);
+          LOG.info("Found expired compaction instant, update the earliest scheduled compaction instant which will be executed: "
+              + cfg.compactionInstantTime);
         }
       }
+
+      // do nothing if cfg.compactionInstantTime still is null
+      if (StringUtils.isNullOrEmpty(cfg.compactionInstantTime)) {
+        LOG.info("There is no scheduled compaction in the table.");
+        return 0;
+      }
+
+      // start a heartbeat for the instant
+      client.getHeartbeatClient().start(cfg.compactionInstantTime);
+
```

Comment:
The heartbeat was previously used for lazy cleaning for multi-writers:

```java
    if (config.getFailedWritesCleanPolicy().isLazy()) {
      this.heartbeatClient.start(instantTime);
    }
```

Now it is used for table service rollback, makes sense to me, we just need to assure the heartbeat client is working during the whole lifecycle of the compaction job.

Reply:
Yes. I forget that the SparkRDDWriteClient is not closed in previous implementation. I add it.

Reply:
The try finally caluse would close the client automically ?

Reply:
It can close the heartbeat client. `client.close` will call `this.heartbeatClient.stop` finally.
```java
  /**
   * Releases any resources used by the client.
   */
  @Override
  public void close() {
    stopEmbeddedServerView(true);
    this.context.setJobStatus("", "");
    this.heartbeatClient.stop();
    this.txnManager.close();
  }
```

Reply:
From the code, I didn't see the difference.

Reply:
I don't particularly understand what you mean. You mean we should call `this.heartbeatClient.stop(instantTime)`  to close the heartbeat client?

Reply:
I mean the original try finally block close the write client after the whole compaction/clean finishes, the new try and close seem does the same thing ?

Reply:
In the original code, it doesn't close the write client. We newly add the finally block to close it.

Reply:
> I mean the original try finally block close the write client after the whole compaction/clean finishes, the new try and close seem does the same thing ?

You are right. I overlooked this knowledge point that the SparkRDDWriteClient implements AutoCloseable. It will close the write client automatically and there is no need to add finally block. But the heartbeat client will also be closed if the write client closed.

Reply:
Did you test it manually offline for multiple compactors?

Reply:
We tested, there are no functional problems for the serial execution and the concurrent execution. But we are concerned about data quality problems for concurrent execution, so we still choose to execute serially now. This pr can satisfy us to rollback the failed inflight instant.